engine/lib/phases/phase_project_instead_of_match.ml

```diff
@@ -0,0 +1,227 @@
+            | None -> (None, [])
+          in
+          ( PBinding
+              {
+                mut = dmutability span S.mutable_variable mut;
+                mode = dbinding_mode span mode;
+                var;
+                typ = dty span typ;
+                subpat = simple_pat;
+                (* TODO *)
+                (* Option.map ~f:(dpat *** S.as_pattern) subpat; *)
+              },
+            remaining_pats )
+      | PDeref { subpat; witness } ->
+          let simple_pat, remaining_pats = project_pat subpat in
+          ( PDeref { subpat = simple_pat; witness = S.reference span witness },
+            remaining_pats )
+      | POr { subpats } ->
+          let updated_subpats = List.map ~f:project_pat subpats in
+          ( POr { subpats = List.map ~f:fst updated_subpats },
+            List.concat_map ~f:snd updated_subpats )
+
+    and let_of_pat_binding ((p, rhs) : B.pat * B.expr) (body : B.expr) : B.expr
+        =
+      UB.make_let p rhs body
+
+    and lets_of_pat_bindings (bindings : (B.pat * B.expr) list) (body : B.expr)
+        : B.expr =
+      List.fold_right ~init:body ~f:let_of_pat_binding bindings
+
+    and dexpr' (span : span) (e : A.expr') : B.expr' =
+      match (UA.unbox_underef_expr { e; span; typ = UA.never_typ }).e with
+      | [%inline_arms "dexpr'.*" - Let - Closure - Loop - Match] -> auto
+      | Match { scrutinee; arms } ->
+          Match { scrutinee = dexpr scrutinee; arms = List.map ~f:darm arms }
+      | Let { monadic; lhs; rhs; body } ->
+          let simple_pat, remaining_pats = project_pat lhs in
+          Let
+            {
+              monadic =
+                Option.map
+                  ~f:(dsupported_monads span *** S.monadic_binding span)
+                  monadic;
+              lhs = simple_pat;
+              rhs = dexpr rhs;
+              body = lets_of_pat_bindings remaining_pats (dexpr body);
+            }
+      | Loop { body; kind; state; label; witness } ->
+          Loop
+            {
+              body = dexpr body;
+              kind = dloop_kind span kind;
+              state = Option.map ~f:(dloop_state span) state;
+              label;
+              witness = S.loop span witness;
+            }
+      | Closure { params; body; captures } ->
+          let projected_params = List.map ~f:project_pat params in
+          Closure
+            {
+              params = List.map ~f:fst projected_params;
+              body =
+                lets_of_pat_bindings
+                  (List.concat_map ~f:snd projected_params)
+                  (dexpr body);
+              captures = List.map ~f:dexpr captures;
+            }
+
+    and darm' (_span : span) (a : A.arm') : B.arm' =
+      let simple_pat, remaining_pats = project_pat a.arm_pat in
+      {
+        arm_pat = simple_pat;
+        body = lets_of_pat_bindings remaining_pats (dexpr a.body);
+      }
+      [@@inline_ands bindings_of dexpr]
```

Comment:
Mmh, I'm a bit worried about that one.
If I understand correctly, here something like:
```rust
match x {
  Some(Foo {a: None, b}) => b,
  Some(Foo {a: Some(a), b}) => a + b,
  None => 0,
}
```
would be rewritten into:
```rust
match x {
  Some(some_name) =>
    match some_name {
      Foo {a: some_name, b} => match some_name { None => b }
    },
  Some(some_name) => 
     match some_name {
       Foo {a: some_name, b} => match some_name { Some(a) => a + b},
     },
  None => 0,
}
```
am I right?

But then, the first sub-sub-match is incomplete and the second arm won't be reached.

Reply:
Currently a similar problem exists for
```rust
#![allow(dead_code)]
pub struct Foo {
    pub a : Option<usize>,
    pub b : usize,
}

pub fn bar(x : Option<Foo>) -> usize {
    match x {
        Some(f) if match f { Foo {a: None, b} => true, _ => false } => {
            let Foo {a: None, b} = f;
            b
        }
        Some(f) if match f { Foo {a: Some(a), b} => true, _ => false } => {
            let Foo {a: Some(a), b} = f;
            a + b
        }
        None => 0,
    }
}
```
which is the correct "dumb" way to do the translation. However I should fix this, such that multiple cases for the same constructor are respected (which is currently not the case).

Reply:
This is indeed an issue, I should have some check about which case I end up in before binding the values.

Reply:
I would expect the translation to look something like:
```rust
match x {
  Some(f) if if let (None, b) = (f.a, f.b) { true } else { false } => {
    let None = f.a;
    let b = f.b;
    b
  }
  None => 0,
  Some(f) if if let (Some(a), b) = (f.a, f.b) { true } else { false } => {
    let Some(a) = f.a;
    let b = f.b;
    a + b
  },
}
```
However we currently do not translate the conditions on match cases?